common/persistence/client/health_request_rate_limiter.go

```diff
@@ -0,0 +1,199 @@
+		RateBackoffStepSize float64
+		// when the system is healthy and current rate < max rate, the current rate multiplier will be
+		// increased by this amount
+		RateIncreaseStepSize float64
+	}
+)
+
+var _ quotas.RequestRateLimiter = (*HealthRequestRateLimiterImpl)(nil)
+
+func NewHealthRequestRateLimiterImpl(
+	healthSignals persistence.HealthSignalAggregator,
+	rateFn quotas.RateFn,
+	params DynamicRateLimitingParams,
+	logger log.Logger,
+) *HealthRequestRateLimiterImpl {
+	limiter := &HealthRequestRateLimiterImpl{
+		enabled:           &atomic.Bool{},
+		rateLimiter:       quotas.NewRateLimiter(rateFn(), int(DefaultRateBurstRatio*rateFn())),
+		healthSignals:     healthSignals,
+		rateFn:            rateFn,
+		params:            params,
+		refreshTimer:      time.NewTicker(DefaultRefreshInterval),
+		rateToBurstRatio:  DefaultRateBurstRatio,
+		minRateMultiplier: DefaultMinRateMultiplier,
+		maxRateMultiplier: DefaultMaxRateMultiplier,
+		curRateMultiplier: DefaultMaxRateMultiplier,
+		logger:            logger,
+	}
+	limiter.refreshDynamicParams()
+	return limiter
+}
+
+func (rl *HealthRequestRateLimiterImpl) Allow(now time.Time, request quotas.Request) bool {
+	rl.maybeRefresh()
+	if !rl.enabled.Load() {
+		return true
+	}
+	return rl.rateLimiter.AllowN(now, request.Token)
+}
+
+func (rl *HealthRequestRateLimiterImpl) Reserve(now time.Time, request quotas.Request) quotas.Reservation {
+	rl.maybeRefresh()
+	if !rl.enabled.Load() {
+		return quotas.NoopReservation
+	}
+	return rl.rateLimiter.ReserveN(now, request.Token)
+}
+
+func (rl *HealthRequestRateLimiterImpl) Wait(ctx context.Context, request quotas.Request) error {
+	rl.maybeRefresh()
+	if !rl.enabled.Load() {
+		return nil
+	}
+	return rl.rateLimiter.WaitN(ctx, request.Token)
+}
+
+func (rl *HealthRequestRateLimiterImpl) maybeRefresh() {
+	select {
+	case <-rl.refreshTimer.C:
+		rl.refreshDynamicParams()
+		if rl.enabled.Load() {
+			rl.refreshRate()
+		}
+		rl.updateRefreshTimer()
+
+	default:
+		// no-op
+	}
+}
+
+func (rl *HealthRequestRateLimiterImpl) refreshRate() {
+	if rl.latencyThresholdExceeded() || rl.errorThresholdExceeded() {
+		// limit exceeded, do backoff
+		rl.curRateMultiplier = math.Max(rl.minRateMultiplier, rl.curRateMultiplier-rl.curOptions.RateBackoffStepSize)
+		rl.rateLimiter.SetRate(rl.curRateMultiplier * rl.rateFn())
+		rl.rateLimiter.SetBurst(int(rl.rateToBurstRatio * rl.rateFn()))
+	} else if rl.curRateMultiplier < rl.maxRateMultiplier {
+		// already doing backoff and under thresholds, increase limit
+		rl.curRateMultiplier = math.Min(rl.maxRateMultiplier, rl.curRateMultiplier+rl.curOptions.RateIncreaseStepSize)
+		rl.rateLimiter.SetRate(rl.curRateMultiplier * rl.rateFn())
+		rl.rateLimiter.SetBurst(int(rl.rateToBurstRatio * rl.rateFn()))
+	}
+}
+
+func (rl *HealthRequestRateLimiterImpl) refreshDynamicParams() {
+	var options dynamicRateLimitingOptions
+	b, err := json.Marshal(rl.params())
```

Comment:
Hmm.. I thought the point of defining the keys explicitly was to avoid the json stuff? So, those constants are only used in the default value and nowhere else?

I would think if you're going to define string constants for the map keys, you should use them to read the map